### PR TITLE
more realistic ball physics

### PR DESCRIPTION
**Pull request by CarlinWilliamson**
[Before](https://photos.app.goo.gl/fzZ2711gMW2VPvNV9)
[After](https://photos.app.goo.gl/skvuqSCD98Xv8o6K7)

increasing `setContactSlip` stops the ball from sticking to the nao's feet while dribbling
changed `BallMass` and `setLinearDrag` so the ball rolls further. similar to videos from 2016/17

### CODE REVIEW

**Comment by CarlinWilliamson** [comment]
I asked for comments about how realistic this ball movement is on slack.

Jayen:
> I'd say your field is too level and too flat to be realistic.  You need some randomness in the direction and deceleration.
> In think the real ball decelerates faster at lower velocities

Ethan:
> I think it looks like the ball is behaving significantly more realistically. The overall distance looks pretty good, though I think in reality it might move slightly farther still.
> I agree that at low velocity the ball decelerates quickly and veers off in a pseudo random direction. In reality this depends on imperfections in the field surface and the mass distribution of the ball, but that's probably pretty out of scope for the simulator. You could consider if it is worth improving it with a simple random direction change.

**Comment by CarlinWilliamson** [comment]
To try and make the ball stop quicker, I have added "Stillness".

It is a "force" which resists the (angular) velocity of the ball. It is calculated by taking an exponential of the negative of the current (angular) velocity, is proportional to the linear and angular drag factors, with a configurable base and scaling factor.
i.e. The force is small while the ball is moving/rolling quickly but ramps up exponentially when the ball is close to stopping

[Before](https://photos.app.goo.gl/9yegPyDJ9g8AthhM6)
[After](https://photos.app.goo.gl/9yegPyDJ9g8AthhM6)

File: spark/lib/oxygen/physicsserver/dragcontroller_c.cpp [hunk 1]
```diff
@@ -65,11 +65,57 @@ FUNCTION(DragController,setAngularDrag)
...
 }
 
+FUNCTION(DragController,getStillness)
+{
+    return obj->GetStillness();
+}
+
+FUNCTION(DragController,setStillness)
+{
+    float inDrag;
```
Comment (jayenashar) [inline spark/lib/oxygen/physicsserver/dragcontroller_c.cpp]: i feel like this should be `inStillness` to decrease confusion

File: spark/lib/oxygen/physicsserver/dragcontroller.cpp [hunk 2]
```diff
@@ -54,6 +56,17 @@ void DragController::PrePhysicsUpdateInternal(float /*deltaTime*/)
         Vector3f vel = mBody->GetAngularVelocity() * mAngularDrag * -1;
         mBody->AddTorque(vel);
     }
+
+    if (mStillness > 0)
+    {
+        // apply linear drag
+        Vector3f vel = mBody->GetVelocity();
+        mBody->AddForce(vel * pow(mStillnessBase, -vel.Length()) * mStillness * mLinearDrag * -1);
```
Comment (jayenashar) [inline spark/lib/oxygen/physicsserver/dragcontroller.cpp]: thanks for not copying the above.  `vel` for drag is confusing

File: spark/lib/oxygen/physicsserver/dragcontroller_c.cpp (hunk 1 above)
Comment (CarlinWilliamson) [reply spark/lib/oxygen/physicsserver/dragcontroller_c.cpp]: Ah. The classic copy pasta XD

Fixed

**Comment by CarlinWilliamson** [comment]
Anyone have suggestions for a better name than stillness? It's not terrible but I'm sure we can do better...

**Comment by jayenashar** [comment]
> Anyone have suggestions for a better name than stillness? It's not terrible but I'm sure we can do better...

`acceleratedDeceleration`?

**Comment by pzrq** [comment]
> > Anyone have suggestions for a better name than stillness? It's not terrible but I'm sure we can do better...
> 
> `acceleratedDeceleration`?

I’d simplify it to `decelerator` though in general find thesaurus useful when I’m close but not convinced I have named my concept accurately, e.g. :
https://www.thesaurus.com/browse/stillness?s=t
https://www.thesaurus.com/browse/decelerate?s=t

Obviously should use inclusive language these days so e.g. retarder is out, alternatives I like are hinderer, inhibitor and restrainer but personally don’t care as long as it’s used consistently once one is chosen.

File: spark/lib/oxygen/physicsserver/dragcontroller.cpp [hunk 3]
```diff
@@ -54,6 +56,17 @@ void DragController::PrePhysicsUpdateInternal(float /*deltaTime*/)
...
+
+        // apply angular drag
+        Vector3f avel = mBody->GetAngularVelocity();
+        mBody->AddTorque(avel * pow(mStillnessBase, -avel.Length()) * mStillness * mAngularDrag * -1);
+    }
```
Comment (pzrq) [inline spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: Perhaps thinking about the torque, velocity and angular drag here may also lead to a better name for stillness?

Comment (jayenashar) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: There's already a drag, so maybe dragExtra?

Comment (pzrq) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: Aside: There's even a [book on this](https://www.namingthings.co/) 😂 - I usually can stop at [Martin Fowler's take on it](https://www.martinfowler.com/bliki/TwoHardThings.html).

Comment (pzrq) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: If in doubt, I'd probably go for more descriptive, so if we can't categorise the type of `mDragExtra` as distinct from `mAngularDrag`, I'd suggest the named equivalents of `mDeceleratorBase` and `mDeceleratorFactor`

Comment (pzrq) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: Though as I reread the formula - any reason that `mStillness` at least (as distinct from `mStillnessBase`) can't be merged into `mAngularDrag`?

Comment (jayenashar) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: if the built-in calculations assume a perfect sphere, and this accounts for the bumps in the ball, we could name it `mDragIrregularBase` and `mDragIrregularFactor`

Comment (jayenashar) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: what do you mean `merged into`?  stillness applies to linear drag as well

Comment (CarlinWilliamson) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: I'm a fan of `mDeceleratorBase` and `mDeceleratorFactor`

Perhaps we should look into making the field a little bumpy or the ball less spherical. Not sure how to atm...

Comment (jayenashar) [reply spark/lib/oxygen/physicsserver/dragcontroller.cpp:69]: it'll make the physics calculations more complicated and the simulator will run slower.  better to add a drag in a random direction of a random amount.

File: rcssserver3d/data/rsg/agent/ball.rsg [hunk 4]
```diff
@@ -20,8 +20,10 @@
 	 
 	 (setSphereTotal $Mass $Radius)
 	 (nd DragController
-	     (setAngularDrag 0.00005)
-	     (setLinearDrag 0.01)
+	     (setAngularDrag 0.00004)
+	     (setLinearDrag 0.018)
+	     (setStillness 1.5)
+	     (setStillnessBase 3)
```
Comment (pzrq) [inline rcssserver3d/data/rsg/agent/ball.rsg]: Still about half a dozen things named stillness.

@CarlinWilliamson When this is ready to go, please re-request a review by clicking on the relevant icon 😄 
Screen Shot 2020-09-16 at 8 32 02 pm

Comment (CarlinWilliamson) [reply rcssserver3d/data/rsg/agent/ball.rsg]: haha. oops. fixed now and re-requested

**Review by pzrq — APPROVED** [review]
LGTM 👍🏼

**Pull request merged**